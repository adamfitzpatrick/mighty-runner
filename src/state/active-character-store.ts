import { observable, reaction, action, ObservableMap, computed } from 'mobx'
import ApiService from '@services/api-service'
import {
  Character,
  Attribute,
  Attributes,
  OtherAttributes,
  Effect,
  PersistableCharacter,
  AttributeName,
  OtherAttributeName
} from '@models'

export interface ActiveCharacterProps {
  activeCharacter?: ActiveCharacterStore
}

/**
 * Class for maintaining the application store for a character being actively
 * viewed or modified by the user.  Static value of the storage key for the local
 * version of the active character is maintained here.  This store carries the side-effect
 * of reading from and storing an active character to the browser's local storage.
 */
export class ActiveCharacterStore {
  static ACTIVE_LOCAL_STORAGE_KEY = 'mighty_runner_character_active'
  static UPDATE_DELAY = 1000

  @observable character: Character | null = null
  @observable attributes: Attributes | null = null
  @observable otherAttributes: OtherAttributes | null = null
  @observable effects: Effect[] | null = null

  constructor () {
    reaction(() => this.character, this.characterReaction)
  }

  /**
   * Action to read an existing active character from local storage if one exists.
   */
  @action
  readFromLocalStorage () {
    let persistableCharacter: PersistableCharacter
    try {
      persistableCharacter = JSON.parse(localStorage.getItem(ActiveCharacterStore.ACTIVE_LOCAL_STORAGE_KEY)!)
      this.extractCharacterObjects(persistableCharacter)
    } catch (e) {
      return
    }
  }

  /**
   * Action which sets the active character from a persistable character object.
   * @param {PersistableCharacter} persistableharacter
   */
  @action
  setActiveCharacter (persistableCharacter: PersistableCharacter) {
    this.persistToLocalStorage(persistableCharacter)
    this.extractCharacterObjects(persistableCharacter)
  }

  /**
   * Action to update a character attribute by replacing it with a new object
   * @param {AttributeName} attributeName name of attribute to be replaced
   * @param {Attribute} attribute replacement attribute object
   */
  @action
  updateAttribute (attributeName: AttributeName, attribute: Attribute) {
    this.attributes![attributeName] = Object.assign({}, attribute)
    this.persistAll()
  }

  /**
   * Action to update a character otherAttribute by replacing it with a new object
   * @param {OtherAttributeName} otherAttributeName name of otherAttribute object to be replaced
   * @param {Attribute} otherAttribute replacement attribute object
   */
  @action
  updateOtherAttribute (otherAttributeName: OtherAttributeName, otherAttribute: Attribute) {
    this.otherAttributes![otherAttributeName] = Object.assign({}, otherAttribute)
    this.persistAll()
  }

  /**
   * Action to update an existing effect on a character by replacing it with a new object
   * @param {Effect} updatedEffect effect with is to replace an existing effect
   */
  @action
  updateEffect (updatedEffect: Effect) {
    const index = this.effects!.findIndex(effect => effect.id === updatedEffect.id)
    // TODO Handle index not found
    const effects = this.effects!.concat()
    effects[index] = updatedEffect
    this.effects = effects
    this.persistAll()
  }
  /**
   * Toggles the state of an existing effect by replacing the effect with a new object
   * whose active state is reversed
   * @param {string} effectId ID string of effect to be toggled
   */
  @action
  toggleEffect (effectId: string) {
    const updated = Object.assign({}, this.effects!.find(effect => effect.id === effectId))
    // TODO Handle index not found
    updated.active = !updated.active
    this.updateEffect(updated)
  }

  /**
   * Adds an effect to the character by replacing the existing array of effects with a new
   * one including the added effect.
   * @param {Effect} effect Effect to be added to the effects array
   */
  @action
  addEffect (effect: Effect) {
    this.effects = this.effects!.concat([ effect ])
    this.persistAll()
  }

  private characterReaction = (): void => {
    if (!this.character) { return }
    this.persistAll()
  }

  private persistAll (): void {
    const persistableCharacter = this.compileCharacterObjects()
    this.persistToLocalStorage(persistableCharacter)
    this.persistToApi(persistableCharacter)
  }

  private persistToLocalStorage (persistableCharacter: PersistableCharacter): void {
    localStorage.setItem(
      ActiveCharacterStore.ACTIVE_LOCAL_STORAGE_KEY,
      JSON.stringify(persistableCharacter)
    )
  }

  private persistToApi (persistableCharacter: PersistableCharacter) {
    ApiService.putCharacter(persistableCharacter.id, persistableCharacter)
      .then(() => { /* no-op */ }, () => { /* no-op */ })
  }

  private extractCharacterObjects (persistableharacter: PersistableCharacter): void {
    this.character = {
      userId: persistableharacter.userId,
      id: persistableharacter.id,
      name: persistableharacter.name
    }
    this.attributes = persistableharacter.attributes
    this.otherAttributes = persistableharacter.otherAttributes
    this.effects = persistableharacter.effects
  }

  private compileCharacterObjects (): PersistableCharacter {
    return {
      userId: this.character!.userId,
      id: this.character!.id,
      name: this.character!.name,
      attributes: this.attributes!,
      otherAttributes: this.otherAttributes!,
      effects: this.effects!
    }
  }
}

export const activeCharacter = new ActiveCharacterStore()
