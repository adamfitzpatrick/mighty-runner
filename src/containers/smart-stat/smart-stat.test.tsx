import { h } from 'preact'
import deep, { RenderContext } from 'preact-render-spy'
import SmartStat from '.'
import { PersistableCharacter, Attribute, Effect } from '@models'
import { activeCharacter } from '@state/active-character-store'

type Props = { statPath: string[], round?: boolean }
type State = { showEffects: boolean }
type EffectsProps = {
  effects: Effect[]
  visible?: boolean
  left?: boolean
  top?: boolean
}

describe('SmartStat container', () => {
  let smartStat: RenderContext<Props, State>
  let persistableCharacter: PersistableCharacter

  beforeEach(() => {
    const character = {
      userId: 'USERID',
      id: '1',
      name: 'Character'
    }
    const attributes = {
      agility: { value: { initial: 1, chargen: 5 }, asTarget: [ 'attributes', 'agility' ] } as Attribute
    }
    const otherAttributes = {
      essence: { value: { initial: 6 }, asTarget: [ 'otherAttributes', 'essence' ] } as Attribute
    }
    const effects = [{
      name: 'Effect 1',
      id: '1',
      value: 1,
      active: true,
      target: [ 'attributes', 'agility' ]
    }, {
      name: 'Effect 2',
      id: '2',
      value: 145,
      active: false,
      target: [ 'otherAttributes', 'essence' ]
    }]
    persistableCharacter = {
      userId: character.userId,
      id: character.id,
      name: character.name,
      attributes,
      otherAttributes,
      effects
    }
    activeCharacter.setActiveCharacter(persistableCharacter)
  })

  beforeEach(() => {
    smartStat = deep(
      <SmartStat
        activeCharacter={activeCharacter}
        statName='agility'
        effectableGroup='attributes'
      />,
      { depth: 3 }
    )
  })

  describe('when rendering', () => {
    test('should show the correct attribute value', () => {
      expect(smartStat).toMatchSnapshot()
      expect(smartStat.find<{ value: string }, {}>('Stat').attrs().value).toEqual(7)
    })

    test('should reflect the activation state of related effects', () => {
      persistableCharacter.effects[0].active = false
      activeCharacter.setActiveCharacter(persistableCharacter)
      smartStat = deep(
        <SmartStat
          activeCharacter={activeCharacter}
          statName='agility'
          effectableGroup='attributes'
        />,
        { depth: 3 }
      )
      expect(smartStat.find<{ value: string }, {}>('Stat').attrs().value).toEqual(6)
    })
  })

  test('should render properly as a round design', () => {
    let smartStat = deep(
      <SmartStat
        activeCharacter={activeCharacter}
        statName='agility'
        effectableGroup='attributes'
        round
      />,
      { depth: 3 }
    )
    expect(smartStat).toMatchSnapshot()
  })

  describe('when clicked', () => {
    test('should expand the effects toggle display with the relevant effects', () => {
      smartStat.find('button').simulate('click')
      let effects = smartStat.find<EffectsProps, {}>('SelectEffects')
      expect(effects.attrs().visible).toBeTruthy()
      expect(effects.attrs().effects.length).toEqual(1)
      smartStat.find('button').simulate('click')
      expect(smartStat.find<EffectsProps, {}>('SelectEffects').attrs().visible).toBeFalsy()
    })
  })

  describe('when click occurs outside of component', () => {
    test('should collapse the effects toggle', (done) => {
      smartStat.find('button').simulate('click')
      expect(smartStat.find<EffectsProps, {}>('SelectEffects').attrs().visible).toBeTruthy()
      document.dispatchEvent(new Event('click'))
      setImmediate(() => {
        expect(smartStat.find<EffectsProps, {}>('SelectEffects').attrs().visible).toBeFalsy()
        done()
      })
    })
  })

  describe('effectClickHandler', () => {
    test('should toggle the appropriate effect state', () => {
      activeCharacter.toggleEffect = jest.fn()
      smartStat.find('EffectToggle').first()[0].attributes.toggleHandler('1')
      expect(activeCharacter.toggleEffect).toHaveBeenCalledWith('1')
    })
  })

  describe('componentDidMount', () => {
    test('should modify the effects display when the stat is near the right or bottom of the screen', done => {
      Element.prototype.getBoundingClientRect = jest.fn(() => {
        return { left: 1000, top: 1000 }
      })
      let smartStat = deep(
        <SmartStat
          activeCharacter={activeCharacter}
          statName='agility'
          effectableGroup='attributes'
          round
        />,
        { depth: 3 }
      )
      setImmediate(() => {
        const effects = smartStat.find<EffectsProps, {}>('SelectEffects')
        expect(effects.attrs().left).toBeTruthy()
        expect(effects.attrs().top).toBeTruthy()
        done()
      })
    })
  })

  describe('componentWillUnmount', () => {
    test('should remove the click event listener from document', () => {
      document.removeEventListener = jest.fn()
      smartStat.render(null as any as JSX.Element)
      expect(document.removeEventListener).toHaveBeenCalled()
    })
  })
})
