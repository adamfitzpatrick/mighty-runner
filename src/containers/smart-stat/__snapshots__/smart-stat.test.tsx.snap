// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmartStat container should render properly as a round design 1`] = `
preact-render-spy (1 nodes)
-------
<div class="smartStat round">
  <button
    class="button"
    onClick={[Function anonymous]}
    aria-label="agility: Select active effects"
  >
    <span class="buttonFocusIndicator"></span>
  </button>
  <div class="stat">
    <div class="stat round">
      <div>7</div>
      <div class="label"></div>
    </div>
  </div>
  <div
    class="effects"
    style="width: 0px;"
  >
    <div class="list">
      <EffectToggle
        effect={
          Object {
            "active": true,
            "id": "1",
            "name": "Effect 1",
            "target": Array [
              "attributes",
              "agility"
            ],
            "value": 1,
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function deepEnhancer$$1],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": 2,
                "name": "ActiveCharacterStore@1.effects[..].keys",
                "observers": Set {}
              },
              "name": "ActiveCharacterStore@1.effects[..]",
              "pendingKeys": Map {
                "toJSON" => false,
                "nodeName" => false
              },
              "proxy": [Circular],
              "target": Object {
                "active": true,
                "id": "1",
                "name": "Effect 1",
                "target": Array [
                  "attributes",
                  "agility"
                ],
                "value": 1,
                Symbol(mobx administration): [Circular]
              },
              "values": Map {
                "name" => "Effect 1",
                "id" => "1",
                "value" => 1,
                "active" => true,
                "target" => Array [
                  "attributes",
                  "agility"
                ]
              }
            }
          }
        }
        toggleHandler={[Function anonymous]}
      >
      </EffectToggle>
    </div>
    <div
      class="editButton"
      style="margin-left: calc(undefinedpx + 2rem);"
    >
      edit stat...
    </div>
  </div>
</div>

`;

exports[`SmartStat container when rendering should show the correct attribute value 1`] = `
preact-render-spy (1 nodes)
-------
<div class="smartStat">
  <button
    class="button"
    onClick={[Function anonymous]}
    aria-label="agility: Select active effects"
  >
    <span class="buttonFocusIndicator"></span>
  </button>
  <div class="stat">
    <div class="stat">
      <div class="label"></div>
      <div>7</div>
    </div>
  </div>
  <div
    class="effects"
    style="width: 0px;"
  >
    <div class="list">
      <EffectToggle
        effect={
          Object {
            "active": true,
            "id": "1",
            "name": "Effect 1",
            "target": Array [
              "attributes",
              "agility"
            ],
            "value": 1,
            Symbol(mobx administration): ObservableObjectAdministration$$1 {
              "defaultEnhancer": [Function deepEnhancer$$1],
              "keysAtom": Atom$$1 {
                "diffValue": 0,
                "isBeingObserved": false,
                "isPendingUnobservation": false,
                "lastAccessedBy": 0,
                "lowestObserverState": 2,
                "name": "ActiveCharacterStore@1.effects[..].keys",
                "observers": Set {}
              },
              "name": "ActiveCharacterStore@1.effects[..]",
              "pendingKeys": Map {
                "toJSON" => false,
                "nodeName" => false
              },
              "proxy": [Circular],
              "target": Object {
                "active": true,
                "id": "1",
                "name": "Effect 1",
                "target": Array [
                  "attributes",
                  "agility"
                ],
                "value": 1,
                Symbol(mobx administration): [Circular]
              },
              "values": Map {
                "name" => "Effect 1",
                "id" => "1",
                "value" => 1,
                "active" => true,
                "target" => Array [
                  "attributes",
                  "agility"
                ]
              }
            }
          }
        }
        toggleHandler={[Function anonymous]}
      >
      </EffectToggle>
    </div>
    <div
      class="editButton"
      style="margin-left: calc(undefinedpx + 2rem);"
    >
      edit stat...
    </div>
  </div>
</div>

`;
